staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -485,27 +623,39 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
 	}
 
 	if oldCRD.UID != newCRD.UID {
-		r.removeStorage_locked(oldCRD.UID)
+		r.removeStorageLocked(oldCRD.UID)
 	}
 
 	storageMap := r.customStorage.Load().(crdStorageMap)
 	oldInfo, found := storageMap[newCRD.UID]
-	if !found {
-		return
-	}
-	if apiequality.Semantic.DeepEqual(&newCRD.Spec, oldInfo.spec) && apiequality.Semantic.DeepEqual(&newCRD.Status.AcceptedNames, oldInfo.acceptedNames) {
-		klog.V(6).Infof("Ignoring customresourcedefinition %s update because neither spec, nor accepted names changed", oldCRD.Name)
-		return
+
+	var tearDownFinishedCh <-chan struct{}
+	if found {
+		if apiequality.Semantic.DeepEqual(&newCRD.Spec, oldInfo.spec) && apiequality.Semantic.DeepEqual(&newCRD.Status.AcceptedNames, oldInfo.acceptedNames) {
+			klog.V(6).Infof("Ignoring customresourcedefinition %s update because neither spec, nor accepted names changed", oldCRD.Name)
+			return
+		}
+		// Tear down the old storage
+		klog.V(4).Infof("Updating customresourcedefinition %s", newCRD.Name)
+		tearDownFinishedCh = r.removeStorageLocked(newCRD.UID)
 	}
 
-	klog.V(4).Infof("Updating customresourcedefinition %s", newCRD.Name)
-	r.removeStorage_locked(newCRD.UID)
+	// Update storage version with the latest info in the watch event
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		ctx := apirequest.NewContext()
+		err := r.storageVersionManager.UpdateStorageVersion(ctx, newCRD, tearDownFinishedCh, nil, nil)
```

Comment:
Working through a potential solution.

Observations
1. Teardowns for ordered CRD updates do NOT complete in order because some requests can be slower than others
2. Because storage version can move between v1 and v2 and back to v1, we should probably avoid writing in order of update and in order of completion because new writes using newer storage versions could invalidate it.
3. Teardown has a maximum time out of two minutes
4. I don't think we really care how quickly the storage version updates.


Idea:
1. What if we create a workqueue for handling StorageVersion writes, keyed by CRD name
2. We can call a "queue CRD for 125 seconds from now" and write into a map[crd.name] -> CRD+timeToWrite
3. when the workqueue pops an item, we lookup in the map, see if timeToWrite has passed and if so, write storage and remove on success (requeue on error).  if not ready,  do nothing


We would not need to have the CRD handle block reads or writes since all old writes would be over by the time we write the storage version.


Some future improvement (that I'd prefer to not have in this PR), could try to track "have all known storage versions completed teardown" and that could be sufficient instead of the two minute timer.

Reply:
Yes this can guarantee that at the time we reach to update the storage version on any CRD, the teardown timeout has expired, but won't there still the race between the timeout expiring and the actual storage being destroyed?

I do feel like this simplifies things a lot due to my suspicion it would remove code waiting and many channel usages. I'm trying to think of a simple way to synchronize things directly

Reply:
Delaying StorageVersion API updates to avoid races between SVM and old version CR writes sounds pragmatic.  It just occurred to me that the there might be a pathological delay when a partitioned apiserver in an HA configuration reconnects to etcd..  so long term it would be ideal if SVM triggers after apiservers agree on storage version, but this seems like a step forward.

Reply:
>  won't there still be a race between the timeout expiring and the actual storage being destroyed?

+1. 

Also IIUC, if we want the new CRD writes to happen only after the old CR writes have completed, we do want to make the CRD writes- blocking or delayed until the previous CR writes have drained? So I dont fully understand why not just use the current approach as it is? Is it because there's a possibility that we might timeout on new CRD writes while waiting for UpdateStorageVersion() to return?

Reply:
Meaning, I dont understand the difference between the approach the PR currently vs. the workqueue approach suggested here. The former blocks a CRD update till the prev CR writes have drained which is guaranteed to finish in 2 mins anyway because of the teardown-timeout, and the latter just delays the CRD update by a fixed amount and relies on the same fact that all requests will either finish or get dropped because of the teardown-timeout. So what;s the advantage of using a workqueue here?

Reply:
After my chat with @deads2k, realized the race condition that the current PR has: because UpdateStorageVersion() is called in 2 places: 
1. within getOrCreateServingInfoFor() which happens when a new CR request comes in 
2. after a CRD update is issued 

The updateSV from #1 does not wait for old CR writes to finish, and thus can result in SV being updated to a newer version while the older CR writes are still in progress. 
As next steps, I think we need to decide :
1. whether we need UpdateStorageVersion() to be called in 2 places. Maybe just calling it after a CRD update suffices?
2. As a part of CRD update, have a way to preserve the status of the SVUpdate (expressed using processedCh, errorCh) which can be referenced at the time of CR writes

Also, another thought on using a workqueue approach as suggested:

> Because storage version can move between v1 and v2 and back to v1, we should probably avoid writing in order of update and in order of completion because new writes using newer storage versions could invalidate it.

There is a concern of missing to write intermediate SV writes in cases when a CRD is updated from v1 -> v2 -> v1 which is expressed [here](https://github.com/kubernetes/kubernetes/pull/96403#discussion_r564306072). Which suggested that we infact need to capture both the v1 -> v2 update and the v2 -> v3 update by UpdatingSV at both the stages.